ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -254,14 +310,15 @@ private void read(ChannelHandlerContext ctx, DataStreamRequestByteBuf request) {
     final StreamInfo info;
     final CompletableFuture<Long> localWrite;
     final List<CompletableFuture<DataStreamReply>> remoteWrites = new ArrayList<>();
+    final StreamMap.Key key = new StreamMap.Key(ctx.channel().id(), request.getStreamId());
     if (isHeader) {
-      info = streams.computeIfAbsent(request.getStreamId(), id -> newStreamInfo(buf));
+      info = streams.computeIfAbsent(key, id -> newStreamInfo(buf));
       localWrite = CompletableFuture.completedFuture(0L);
       for (DataStreamOutput out : info.getDataStreamOutputs()) {
         remoteWrites.add(out.getHeaderFuture());
       }
     } else {
-      info = streams.get(request.getStreamId());
+      info = streams.get(key);
       localWrite = info.getStream().thenApply(stream -> writeTo(buf, stream));
```

Comment:
@szetszwo Just a discussion, not related to this PR. Because the future `info.getStream()` has completed when stream data, so writeTo was called synchronizedly, do you think writeTo will slow down the multi-stream? Because multi-stream use the same thread to read if there is only one client.

Reply:
> ... so writeTo was called synchronizedly ...

It is still called asynchronously. It said it will use "default execution" in the javadoc https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html .  Indeed it is a good idea to change all the calls to thenApplyAsync with an Executor parameter.  Do you want to reopen https://issues.apache.org/jira/browse/RATIS-1122 and work on it?

Reply:
@szetszwo I think the javadoc is not clear about this. You can find the details [here](https://4comprehension.com/completablefuture-the-difference-between-thenapply-thenapplyasync/) 
![image](https://user-images.githubusercontent.com/51938049/97773397-5d51a280-1b8a-11eb-8926-9e3dc3ecb307.png).
I also test it, it's not called asynchronously.

Reply:
You are right.  I also have tested it.  Then, we should change it to use the async version.

Reply:
@szetszwo Thanks. I reopen RATIS-1122 and will work on it after I finish https://issues.apache.org/jira/browse/RATIS-1083.

Reply:
Sure, thanks.